x-pack/apm-server/main.go

```diff
@@ -178,6 +178,8 @@ func runServerWithProcessors(ctx context.Context, runServer beater.RunServerFunc
 	runServer = beater.WrapRunServerWithProcessors(runServer, batchProcessors...)
 
 	g, ctx := errgroup.WithContext(ctx)
+	serverctx, cancel := context.WithCancel(context.Background())
+	ctx = context.WithValue(ctx, beater.ServerStopped{}, cancel)
```

Comment:
```suggestion
	serverStopped := make(chan struct{})
```

In the processor.Stop goroutine:

```go
		g.Go(func() error {
			<-serverStopped
			stopctx := context.Background()
```

In the runServer goroutine:

```go
	g.Go(func() error {
		defer close(serverStopped)
		return runServer(ctx, args)
	})
```

Reply:
that's much better